chain/ethereum/src/ethereum_adapter.rs

```diff
@@ -288,68 +284,65 @@ where
-
-                            // If the step is already 0, we're hitting the log
-                            // limit even for a single block. We hope this never
-                            // happens, but if it does, make sure to error.
-                            if TOO_MANY_LOGS_FINGERPRINTS
-                                .iter()
-                                .any(|f| string_err.contains(f))
-                                && step > 0
-                            {
-                                // The range size for a request is `step + 1`.
-                                // So it's ok if the step goes down to 0, in
-                                // that case we'll request one block at a time.
-                                let new_step = step / 10;
-                                debug!(logger, "Reducing block range size to scan for events";
+                let res = eth
+                    .logs_with_sigs(
+                        &logger,
+                        subgraph_metrics.clone(),
+                        start,
+                        end,
+                        filter.clone(),
+                        TOO_MANY_LOGS_FINGERPRINTS,
+                    )
+                    .compat()
+                    .await;
+
+                match res {
+                    Err(e) => {
+                        let string_err = e.to_string();
+
+                        // If the step is already 0, the request is too heavy even for a single
+                        // block. We hope this never happens, but if it does, make sure to error.
+                        if TOO_MANY_LOGS_FINGERPRINTS
+                            .iter()
+                            .any(|f| string_err.contains(f))
+                            && step > 0
+                        {
+                            // The range size for a request is `step + 1`. So it's ok if the step
+                            // goes down to 0, in that case we'll request one block at a time.
+                            let new_step = step / 10;
+                            debug!(logger, "Reducing block range size to scan for events";
                                                "new_size" => new_step + 1);
-                                Ok((vec![], (start, new_step)))
-                            } else {
-                                warn!(logger, "Unexpected RPC error"; "error" => &string_err);
-                                Err(err_msg(string_err))
-                            }
+                            Ok(Some((vec![], (start, new_step))))
+                        } else {
+                            warn!(logger, "Unexpected RPC error"; "error" => &string_err);
+                            Err(err_msg(string_err))
                         }
-                        Ok(logs) => Ok((logs, (low, step))),
-                    }),
-            )
+                    }
+                    Ok(logs) => Ok(Some((logs, (end + 1, step)))),
```

Comment:
Is the nature of asking for too many logs that there are simply too many in that range, but that a future range may contain fewer logs and successfully complete with a larger step?

If so, I would consider increasing the step size on successful requests. Eg: `Ok(Some((logs, (end + 1, max_step.min(step * 2)))))`. As the code is, the step size can (and will) only decrease the step size, slowing down and never recovering the speed. The `max_step` variable would simply be the value set earlier at `let step = match filter.contracts.is_empty() {`... This value would be both `max_step` and `step`.

Reply:
I'm not convinced that it's simply the number of logs in all cases, the Alchemy rinkeby nodes where returning -32000 at any range over 500 blocks, even for the first blocks of the network. Hopefully this is something they can improve on their side.

For Infura it really is about the number of logs, but limit is now 10,000 logs which is well above our target triggers per range (currently 100 by default), so this should be a rare event anyways.

So I don't think it's worth having logic that tries to find the optimum range here, that's kinda what our adaptive block range already does, this is just a fail safe.

Reply:
Ok.